### ISSUE TITLE
Get last activities and create a calendar event for them

### ISSUE THREAD

**Issue by victor-falcon** [RafhaanShah/TaskerHealthConnect#2]
I want to create a task to get last activities from health connect and create a calendar event for them.

If I'm not wrong right now, it's impossible with this plugging as it is. 

It is difficult right now or an easy update that we can make? 🤔

**Comment by RafhaanShah** [comment]
Hello, there's a lot more new data available in [0.0.2](https://github.com/RafhaanShah/TaskerHealthConnect/releases/tag/0.0.2) so this may be enough for your needs

**Comment by strazto** [comment]
Btw - I made a scheduled task that pulls SleepRecords and pushes them to calendar, sharing as XML below

```xml
<TaskerData sr="" dvi="1" tv="6.5.11">
	<Profile sr="prof12" ve="2">
		<cdate>1751851250256</cdate>
		<edate>1751851250256</edate>
		<flags>8</flags>
		<id>12</id>
		<mid0>2</mid0>
		<Time sr="con0">
			<fh>19</fh>
			<fm>0</fm>
			<th>19</th>
			<tm>0</tm>
		</Time>
	</Profile>
	<Project sr="proj0" ve="2">
		<cdate>1751700505271</cdate>
		<name>HealthConnect Sleep to Calendar</name>
		<pids>12</pids>
		<tids>2</tids>
		<Share sr="Share">
			<b>false</b>
			<d>Writes HealthConnect sleep data for the day to a calendar named "Sleep".

writes all records for the past 24 hours every evening at 8pm.

requires tasker healthconnect plugin and CalendarTask plugin.</d>
			<g>Calendar</g>
			<p>true</p>
			<t></t>
		</Share>
	</Project>
	<Task sr="task2">
		<cdate>1751700530043</cdate>
		<edate>1751851112036</edate>
		<id>2</id>
		<nme>Pull HealthConnect</nme>
		<pri>100</pri>
		<Action sr="act0" ve="7">
			<code>547</code>
			<label>set MS in a day</label>
			<Str sr="arg0" ve="3">%msInDay</Str>
			<Str sr="arg1" ve="3">24*60*60*1000</Str>
			<Int sr="arg2" val="0"/>
			<Int sr="arg3" val="1"/>
			<Int sr="arg4" val="0"/>
			<Int sr="arg5" val="3"/>
			<Int sr="arg6" val="1"/>
		</Action>
		<Action sr="act1" ve="7">
			<code>547</code>
			<label>get time yesterday in MS</label>
			<Str sr="arg0" ve="3">%timeYesterday</Str>
			<Str sr="arg1" ve="3">%TIMEMS - %msInDay</Str>
			<Int sr="arg2" val="0"/>
			<Int sr="arg3" val="1"/>
			<Int sr="arg4" val="0"/>
			<Int sr="arg5" val="3"/>
			<Int sr="arg6" val="1"/>
		</Action>
		<Action sr="act2" ve="7">
			<code>1187787926</code>
			<Bundle sr="arg0">
				<Vals sr="val">
					<com.twofortyfouram.locale.intent.extra.BLURB>End Timestamp: %TIMEMS
Class of the type of Health Record: SleepSessionRecord
Start Timestamp: %timeYesterday</com.twofortyfouram.locale.intent.extra.BLURB>
					<com.twofortyfouram.locale.intent.extra.BLURB-type>java.lang.String</com.twofortyfouram.locale.intent.extra.BLURB-type>
					<endTime>%TIMEMS</endTime>
					<endTime-type>java.lang.String</endTime-type>
					<net.dinglisch.android.tasker.RELEVANT_VARIABLES>&lt;StringArray sr=""&gt;&lt;_array_net.dinglisch.android.tasker.RELEVANT_VARIABLES0&gt;%healthconnectresult
Read Data
Data will be returned as a JSON object, see HealthConnectClient.readRecords API for the response format.&lt;/_array_net.dinglisch.android.tasker.RELEVANT_VARIABLES0&gt;&lt;_array_net.dinglisch.android.tasker.RELEVANT_VARIABLES1&gt;%err
Error Code
If there's an error, contains an error code&lt;/_array_net.dinglisch.android.tasker.RELEVANT_VARIABLES1&gt;&lt;_array_net.dinglisch.android.tasker.RELEVANT_VARIABLES2&gt;%errmsg
Error Message
If there's an error, contains an error message&lt;/_array_net.dinglisch.android.tasker.RELEVANT_VARIABLES2&gt;&lt;/StringArray&gt;</net.dinglisch.android.tasker.RELEVANT_VARIABLES>
					<net.dinglisch.android.tasker.RELEVANT_VARIABLES-type>[Ljava.lang.String;</net.dinglisch.android.tasker.RELEVANT_VARIABLES-type>
					<net.dinglisch.android.tasker.extras.ACTION_INPUT_CLASS>com.rafapps.taskerhealthconnect.read.ReadDataInput</net.dinglisch.android.tasker.extras.ACTION_INPUT_CLASS>
					<net.dinglisch.android.tasker.extras.ACTION_INPUT_CLASS-type>java.lang.String</net.dinglisch.android.tasker.extras.ACTION_INPUT_CLASS-type>
					<net.dinglisch.android.tasker.extras.ACTION_RUNNER_CLASS>com.rafapps.taskerhealthconnect.read.ReadDataActionRunner</net.dinglisch.android.tasker.extras.ACTION_RUNNER_CLASS>
					<net.dinglisch.android.tasker.extras.ACTION_RUNNER_CLASS-type>java.lang.String</net.dinglisch.android.tasker.extras.ACTION_RUNNER_CLASS-type>
					<net.dinglisch.android.tasker.extras.EXTRA_WAS_CONFIGURED_BEFORE>true</net.dinglisch.android.tasker.extras.EXTRA_WAS_CONFIGURED_BEFORE>
					<net.dinglisch.android.tasker.extras.EXTRA_WAS_CONFIGURED_BEFORE-type>java.lang.Boolean</net.dinglisch.android.tasker.extras.EXTRA_WAS_CONFIGURED_BEFORE-type>
					<net.dinglisch.android.tasker.extras.VARIABLE_REPLACE_KEYS>endTime recordType startTime</net.dinglisch.android.tasker.extras.VARIABLE_REPLACE_KEYS>
					<net.dinglisch.android.tasker.extras.VARIABLE_REPLACE_KEYS-type>java.lang.String</net.dinglisch.android.tasker.extras.VARIABLE_REPLACE_KEYS-type>
					<net.dinglisch.android.tasker.subbundled>true</net.dinglisch.android.tasker.subbundled>
					<net.dinglisch.android.tasker.subbundled-type>java.lang.Boolean</net.dinglisch.android.tasker.subbundled-type>
					<recordType>SleepSessionRecord</recordType>
					<recordType-type>java.lang.String</recordType-type>
					<startTime>%timeYesterday</startTime>
					<startTime-type>java.lang.String</startTime-type>
				</Vals>
			</Bundle>
			<Str sr="arg1" ve="3">com.rafapps.taskerhealthconnect</Str>
			<Str sr="arg2" ve="3">com.rafapps.taskerhealthconnect.read.ReadDataActivity</Str>
			<Int sr="arg3" val="60"/>
			<Int sr="arg4" val="1"/>
		</Action>
		<Action sr="act3" ve="7">
			<code>548</code>
			<on>false</on>
			<Str sr="arg0" ve="3">%healthconnectresult</Str>
			<Int sr="arg1" val="1"/>
			<Str sr="arg10" ve="3"/>
			<Int sr="arg11" val="1"/>
			<Int sr="arg12" val="0"/>
			<Str sr="arg13" ve="3"/>
			<Int sr="arg14" val="0"/>
			<Str sr="arg15" ve="3"/>
			<Int sr="arg2" val="1"/>
			<Str sr="arg3" ve="3"/>
			<Str sr="arg4" ve="3"/>
			<Str sr="arg5" ve="3"/>
			<Str sr="arg6" ve="3"/>
			<Str sr="arg7" ve="3"/>
			<Str sr="arg8" ve="3"/>
			<Int sr="arg9" val="1"/>
		</Action>
		<Action sr="act4" ve="7">
			<code>39</code>
			<Str sr="arg0" ve="3">%record</Str>
			<Str sr="arg1" ve="3">%healthconnectresult.records()</Str>
			<Int sr="arg2" val="1"/>
		</Action>
		<Action sr="act5" ve="7">
			<code>547</code>
			<Str sr="arg0" ve="3">%startUtcSeconds</Str>
			<Str sr="arg1" ve="3">%record.startTime / 1000</Str>
			<Int sr="arg2" val="0"/>
			<Int sr="arg3" val="1"/>
			<Int sr="arg4" val="0"/>
			<Int sr="arg5" val="0"/>
			<Int sr="arg6" val="1"/>
		</Action>
		<Action sr="act6" ve="7">
			<code>547</code>
			<Str sr="arg0" ve="3">%endUtcSeconds</Str>
			<Str sr="arg1" ve="3">%record.endTime / 1000</Str>
			<Int sr="arg2" val="0"/>
			<Int sr="arg3" val="1"/>
			<Int sr="arg4" val="0"/>
			<Int sr="arg5" val="0"/>
			<Int sr="arg6" val="1"/>
		</Action>
		<Action sr="act7" ve="7">
			<code>548</code>
			<Str sr="arg0" ve="3">%endUtcSeconds %startUtcSeconds</Str>
			<Int sr="arg1" val="0"/>
			<Str sr="arg10" ve="3"/>
			<Int sr="arg11" val="1"/>
			<Int sr="arg12" val="0"/>
			<Str sr="arg13" ve="3"/>
			<Int sr="arg14" val="0"/>
			<Str sr="arg15" ve="3"/>
			<Int sr="arg2" val="0"/>
			<Str sr="arg3" ve="3"/>
			<Str sr="arg4" ve="3"/>
			<Str sr="arg5" ve="3"/>
			<Str sr="arg6" ve="3"/>
			<Str sr="arg7" ve="3"/>
			<Str sr="arg8" ve="3"/>
			<Int sr="arg9" val="1"/>
		</Action>
		<Action sr="act8" ve="7">
			<code>764435524</code>
			<Bundle sr="arg0">
				<Vals sr="val">
					<com.balda.calendartask.extra.CALENDAR>26</com.balda.calendartask.extra.CALENDAR>
					<com.balda.calendartask.extra.CALENDAR-type>java.lang.String</com.balda.calendartask.extra.CALENDAR-type>
					<com.balda.calendartask.extra.DESCRIPTION># Notes

%healthconnectresult.records.notes</com.balda.calendartask.extra.DESCRIPTION>
					<com.balda.calendartask.extra.DESCRIPTION-type>java.lang.String</com.balda.calendartask.extra.DESCRIPTION-type>
					<com.balda.calendartask.extra.END>%endUtcSeconds</com.balda.calendartask.extra.END>
					<com.balda.calendartask.extra.END-type>java.lang.String</com.balda.calendartask.extra.END-type>
					<com.balda.calendartask.extra.INT_VERSION_CODE>67</com.balda.calendartask.extra.INT_VERSION_CODE>
					<com.balda.calendartask.extra.INT_VERSION_CODE-type>java.lang.Integer</com.balda.calendartask.extra.INT_VERSION_CODE-type>
					<com.balda.calendartask.extra.OPERATION>1</com.balda.calendartask.extra.OPERATION>
					<com.balda.calendartask.extra.OPERATION-type>java.lang.Integer</com.balda.calendartask.extra.OPERATION-type>
					<com.balda.calendartask.extra.REM_METHOD>-1</com.balda.calendartask.extra.REM_METHOD>
					<com.balda.calendartask.extra.REM_METHOD-type>java.lang.Integer</com.balda.calendartask.extra.REM_METHOD-type>
					<com.balda.calendartask.extra.START>%startUtcSeconds</com.balda.calendartask.extra.START>
					<com.balda.calendartask.extra.START-type>java.lang.String</com.balda.calendartask.extra.START-type>
					<com.balda.calendartask.extra.TIMEZONE>Australia/Sydney</com.balda.calendartask.extra.TIMEZONE>
					<com.balda.calendartask.extra.TIMEZONE-type>java.lang.String</com.balda.calendartask.extra.TIMEZONE-type>
					<com.balda.calendartask.extra.TITLE>Sleep %healthconnectresult.records.title</com.balda.calendartask.extra.TITLE>
					<com.balda.calendartask.extra.TITLE-type>java.lang.String</com.balda.calendartask.extra.TITLE-type>
					<com.twofortyfouram.locale.intent.extra.BLURB>Add Sleep %healthconnectresult.records.title to Sleep</com.twofortyfouram.locale.intent.extra.BLURB>
					<com.twofortyfouram.locale.intent.extra.BLURB-type>java.lang.String</com.twofortyfouram.locale.intent.extra.BLURB-type>
					<net.dinglisch.android.tasker.RELEVANT_VARIABLES>&lt;StringArray sr=""&gt;&lt;_array_net.dinglisch.android.tasker.RELEVANT_VARIABLES0&gt;%ctevid
Event id
&lt;/_array_net.dinglisch.android.tasker.RELEVANT_VARIABLES0&gt;&lt;/StringArray&gt;</net.dinglisch.android.tasker.RELEVANT_VARIABLES>
					<net.dinglisch.android.tasker.RELEVANT_VARIABLES-type>[Ljava.lang.String;</net.dinglisch.android.tasker.RELEVANT_VARIABLES-type>
					<net.dinglisch.android.tasker.extras.VARIABLE_REPLACE_KEYS>com.balda.calendartask.extra.DESCRIPTION com.balda.calendartask.extra.TITLE com.balda.calendartask.extra.TIMEZONE com.balda.calendartask.extra.END com.balda.calendartask.extra.START</net.dinglisch.android.tasker.extras.VARIABLE_REPLACE_KEYS>
					<net.dinglisch.android.tasker.extras.VARIABLE_REPLACE_KEYS-type>java.lang.String</net.dinglisch.android.tasker.extras.VARIABLE_REPLACE_KEYS-type>
					<net.dinglisch.android.tasker.subbundled>true</net.dinglisch.android.tasker.subbundled>
					<net.dinglisch.android.tasker.subbundled-type>java.lang.Boolean</net.dinglisch.android.tasker.subbundled-type>
				</Vals>
			</Bundle>
			<Str sr="arg1" ve="3">com.balda.calendartask</Str>
			<Str sr="arg2" ve="3">com.balda.calendartask.ui.FireAddEvent</Str>
			<Int sr="arg3" val="30"/>
			<Int sr="arg4" val="1"/>
		</Action>
		<Action sr="act9" ve="7">
			<code>40</code>
		</Action>
	</Task>
</TaskerData>
```

Requires

the CalendarTask plugin: `com.balda.calendartask` and of course  this plugin,`com.rafapps.taskerhealthconnect`.

Technically you could do this without CalendarTask, but the native tasker calendar API only lets you add events relative the current time, and I didn't like that API (though it wouldnt exactly be hard to calculate the difference)
